File: PSVRPlayer/Handy/FunscriptPlayer.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text.Json;

namespace PSVRPlayer.Handy;

/// <summary>Funscript parser and per-frame position interpolator.</summary>
public sealed class FunscriptPlayer
{
    public record Action(long At, int Pos); // At = milliseconds, Pos = 0-100

    private List<Action> _actions = new();
    public  bool Loaded => _actions.Count > 0;
    public  IReadOnlyList<Action> Actions => _actions;
    public  long DurationMs => _actions.Count > 0 ? _actions[^1].At : 0;

    public void Load(string json)
    {
        using var doc = JsonDocument.Parse(json);
        var root = doc.RootElement;

        if (!root.TryGetProperty("actions", out var arr))
            throw new FormatException("actionsフィールドが見つかりません");

        _actions = arr.EnumerateArray()
            .Select(e => new Action(e.GetProperty("at").GetInt64(), e.GetProperty("pos").GetInt32()))
            .OrderBy(a => a.At)
            .ToList();
    }

    public void Clear() => _actions.Clear();

    /// <summary>Linear-interpolated position (0-100) at <paramref name="timeMsec"/>.</summary>
    public int GetPositionAt(long timeMsec)
    {
        if (_actions.Count == 0) return 0;
        if (timeMsec <= _actions[0].At)  return _actions[0].Pos;
        if (timeMsec >= _actions[^1].At) return _actions[^1].Pos;

        int lo = 0, hi = _actions.Count - 1;
        while (lo < hi - 1)
        {
            int mid = (lo + hi) >> 1;
            if (_actions[mid].At <= timeMsec) lo = mid; else hi = mid;
        }

        var a = _actions[lo]; var b = _actions[hi];
        double t = (double)(timeMsec - a.At) / (b.At - a.At);
        return (int)Math.Round(a.Pos + (b.Pos - a.Pos) * t);
    }
}

File: src/FunscriptPlayer.js
/**
 * Funscript parser and local position calculator.
 * Funscript format: { version, actions: [{ at: ms, pos: 0-100 }, ...] }
 */
export class FunscriptPlayer {
  constructor() {
    this.actions = [];
    this.loaded = false;
  }

  load(jsonData) {
    const data = typeof jsonData === 'string' ? JSON.parse(jsonData) : jsonData;

    if (!data.actions || !Array.isArray(data.actions)) {
      throw new Error('Invalid funscript: missing actions array');
    }

    // Sort by time ascending
    this.actions = [...data.actions].sort((a, b) => a.at - b.at);
    this.loaded = true;

    return {
      duration: this.actions.length > 0 ? this.actions[this.actions.length - 1].at : 0,
      actionCount: this.actions.length,
    };
  }

  clear() {
    this.actions = [];
    this.loaded = false;
  }

  /**
   * Get interpolated position (0-100) at given time (milliseconds).
   */
  getPositionAt(timeMsec) {
    if (!this.loaded || this.actions.length === 0) return 0;

    const actions = this.actions;

    if (timeMsec <= actions[0].at) return actions[0].pos;
    if (timeMsec >= actions[actions.length - 1].at) return actions[actions.length - 1].pos;

    // Binary search for surrounding keyframes
    let lo = 0;
    let hi = actions.length - 1;
    while (lo < hi - 1) {
      const mid = (lo + hi) >> 1;
      if (actions[mid].at <= timeMsec) lo = mid;
      else hi = mid;
    }

    const a = actions[lo];
    const b = actions[hi];
    const t = (timeMsec - a.at) / (b.at - a.at);
    return Math.round(a.pos + (b.pos - a.pos) * t);
  }
}
